### PR TITLE
MRG: Proximal Policy Optimization (PPO)

### PR DESCRIPTION
**Pull request by smandic5**
PPO algorithm. Probably needs some adjustments so that the code style matches the rest of rl_blox.

### CODE REVIEW

File: rl_blox/algorithm/ppo.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        tuple[jax.Array, jax.Array]: Selected action and it's probability
+    """
+    logits = actor(obs)
+    probs = jax.nn.softmax(logits)
+    action = jax.random.categorical(key, logits)
+    logp = jnp.log(probs[action])
+    return namedtuple('SelectedAction', ['action', 'logp'])(action, logp)
+
+@jax.jit
+def compute_gae(
```
[the remark this answers is not in the document]
Comment (smandic5) [reply rl_blox/algorithm/ppo.py]: Should it be its own file or added to an existing one?

File: rl_blox/algorithm/ppo.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    entropy = -jnp.mean(jnp.sum(probs * jnp.log(probs + 1e-8), axis=-1))
+    return policy_loss + 0.5 * value_loss - 0.01 * entropy
+
+def collect_trajectories(
+    env: gym.Env,
+    actor: nnx.Module,
+    critic: nnx.Module,
+    key: jax.Array,
+    batch_size: int=64
+) -> tuple[jax.Array, jax.Array, jax.Array, jax.Array, jax.Array, jax.Array]:
```
[the remark this answers is not in the document]
Comment (smandic5) [reply rl_blox/algorithm/ppo.py]: Doesn't seem like that is the case, as the dataset doesn't include the terminated flag, and it doesn't include some values like log probabilities, which would have to be calculated again.

File: rl_blox/algorithm/ppo.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+    Returns
+    -------
+    action : jnp.ndarray
+        Selected action.
+    logp : jnp.ndarray
+        Log-probability of the selected action.
+    """
+    logits = actor(obs)
+    probs = jax.nn.softmax(logits)
```
Comment (AlexanderFabisch) [inline rl_blox/algorithm/ppo.py]: Can you solve this with a [softmax policy head](https://github.com/mlaux1/rl-blox/blob/main/rl_blox/blox/function_approximator/policy_head.py#L226)?

File: rl_blox/algorithm/ppo.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    """
+    key = jax.random.key(seed)
+    last_observation, _ = env.reset(seed=seed)
+
+    for episode in trange(episodes, disable=not progress_bar):
+        key, subkey = jax.random.split(key)
+        observations, actions, logps, rewards, terminateds, next_values, last_observation = collect_trajectories(
+            env, actor, critic, subkey, batch_size, last_observation)
+
+        loss_val, grad_actor, grad_critic = calculate_ppo_grads(
```
Comment (AlexanderFabisch) [inline rl_blox/algorithm/ppo.py]: I believe there is no need to have a function for the purpose of calculating the gradients. You can do the optimization step in the same function. Then we have everything jitted.

File: rl_blox/blox/gae.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import jax
+import jax.numpy as jnp
+from collections import namedtuple
+
```
[the remark this answers is not in the document]
Comment (mlaux1) [reply rl_blox/blox/gae.py]: Good point. I recommend setting up [pre-commit](https://pre-commit.com/) as described in the contribution guide (see #186) to help keeping code style consistent.
